src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -570,16 +570,11 @@ private String key() {
 
     /** Handles normal build completion (including errors) but also handles the case that the flow did not even start correctly, for example due to an error in {@link FlowExecution#start}. */
     private void finish(@Nonnull Result r, @CheckForNull Throwable t) {
-        boolean nullListener = false;
-        synchronized (getLogCopyGuard()) {
-            nullListener = listener == null;
+        try {
             setResult(r);
```

Comment:
Once `result` is set to a non-null value, `isBuilding` will call [`isInProgress`](https://github.com/jenkinsci/workflow-job-plugin/blob/db692f18de005fe84d5a95b0bf74597fc8a201a8/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L746) instead of short-circuiting. Doesn't this mean that the behavior will be the same since `FlowExecution.isComplete() == true` (or is it false here?), so `isInProgress` just starts returning false out of the second branch rather than the first due to moving `completed` around?

A good downstream check might be to remove the workaround [here](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/31569c0385f555f0b3146c68eb748a5dcf16efd6/pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/AbstractModelDefTest.java#L511-L539) and run pipeline-model-definition's tests a few times to see if we still run into related failures.

Reply:
> Doesn't this mean

Unfortunately it does. Actually this could happen regardless of when we call `setResult` in `finish`, since this _may_ be called earlier on a truly running build (e.g., by `junit`). So I will need to do deeper fixes.

> remove the workaround [here](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/31569c0385f555f0b3146c68eb748a5dcf16efd6/pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/AbstractModelDefTest.java#L511-L539)

Thanks for tip, I did not know about this one.

Reply:
I checked the comment in `pipeline-model-definition`. I believe that situation is unaffected by this patch:

* `waitForCompletion` starts waiting
* `result` is set to something in the middle of the build, say `UNSTABLE`
* `FlowEndNode` is added to the graph
* (asynch) `GraphListener`s start to be notified
* `isBuilding` is checked
  * `result == null` is false, so `isInProgress()` is called
  * `completed` is still false
  * `exec.isComplete()` is true
* so `assertBuildStatus` fails
* now `GraphL` is notified and `finish` is called
  * in some order, `result` is updated, `completed` is set to `true`, etc.

Fixing that kind of failure would require a more intrusive patch to `isBuilding` which is not required for the originally reported class of flakes and which I am not attempting here.

Reply:
…though you could probably do something like

```java
j.waitForCompletion(run);
j.waitForMessage("Finished: FAILURE", run);
```

which would be reliable after this patch.

Not going to try it so long as https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/325 remains unmerged.

Reply:
> Fixing that kind of failure would require a more intrusive patch to isBuilding which is not required for the originally reported class of flakes and which I am not attempting here

Fair enough. My first thought is to add a `transient volatile boolean currentlyCompleting` field, set it to true before calling `setResult()` or setting `completed = true` in `finish` (set it back to false in a finally block before notifying listeners in `finish`), and then in `isInProgress`, if that field is true, return true (if `finish` can be called multiple times, you'd need something more complicated so a finished build couldn't go back to being in-progress, although it seems unlikely that anyone would be polling `isBuilding` in that case).

Reply:
Would not work unless you registered a `GraphListener.Synchronous` which set this field.

Simpler would be to make `isBuilding` check, say, `listener == null` like `isLogUpdated` now does, but I am not confident in such a change. Of course all these methods are only scantily documented in `Run`, since their original meaning was heavily tied to `AbstractBuild` semantics.